cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/MetadataStoreDataset.java

```diff
@@ -157,42 +156,87 @@ public boolean exists(MDSKey id) {
   // returns mapping of all that has first id parts in range of startId and stopId
   public <T> Map<MDSKey, T> listKV(MDSKey startId, @Nullable MDSKey stopId, Type typeOfT, int limit,
                                    Predicate<T> filter) {
+    return listKV(startId, stopId, typeOfT, limit, null, filter);
+  }
+
+  public <T> Map<MDSKey, T> listKV(MDSKey startId, @Nullable MDSKey stopId, Type typeOfT, int limit,
+                                   Predicate<MDSKey> keyFilter, Predicate<T> valueFilter) {
     byte[] startKey = startId.getKey();
     byte[] stopKey = stopId == null ? Bytes.stopKeyForPrefix(startKey) : stopId.getKey();
 
     Scan scan = new Scan(startKey, stopKey);
-    return listKV(scan, typeOfT, limit, filter);
+    return listKV(scan, typeOfT, limit, keyFilter, valueFilter);
   }
 
-  // returns mapping of all that has first id parts in range of startId and stopId
-  private <T> Map<MDSKey, T> listKV(Scan runScan, Type typeOfT, int limit, Predicate<T> filter) {
+  private <T> Map<MDSKey, T> listCombinedFilterKV(Scan runScan, Type typeOfT, int limit,
+                                                  @Nullable Predicate<KeyValue<T>> combinedFilter) {
     try {
       Map<MDSKey, T> map = Maps.newLinkedHashMap();
-      Scanner scan = table.scan(runScan);
-      try {
+      try (Scanner scan = table.scan(runScan)) {
         Row next;
         while ((limit > 0) && (next = scan.next()) != null) {
+          MDSKey key = new MDSKey(next.getRow());
           byte[] columnValue = next.get(COLUMN);
           if (columnValue == null) {
             continue;
           }
           T value = deserialize(columnValue, typeOfT);
 
-          if (filter.apply(value)) {
-            MDSKey key = new MDSKey(next.getRow());
-            map.put(key, value);
-            --limit;
+          KeyValue<T> kv = new KeyValue<>(key, value);
+          // Combined Filter doesn't pass
+          if (combinedFilter != null && !combinedFilter.apply(kv)) {
+            continue;
           }
+
+          map.put(kv.getKey(), kv.getValue());
+          limit--;
+        }
+        return map;
+      }
+    } catch (Exception e) {
+      throw Throwables.propagate(e);
+    }
+  }
+
+  private <T> Map<MDSKey, T> listKV(Scan runScan, Type typeOfT, int limit, @Nullable Predicate<MDSKey> keyFilter,
```

Comment:
It would be good to consolidate this method into the previous one. Since it is not critical, you could do it later too.